### ISSUE TITLE
mpmc: port_gone_concurrent: test is unsound on single-core systems

### ISSUE THREAD

**Issue by xobs** [rust-lang/rust#107364]
I'm porting libstd to `riscv32imac-unknown-xous-elf`, and as part of the merge effort I've gotten libstd tests running.

This platform has one CPU core and a process has around 8 MB of RAM. Right now, `port_gone_concurrent` is failing:

```rust
#[test]
fn port_gone_concurrent() {
    let (tx, rx) = channel::<i32>();
    let _t = thread::spawn(move || {
        rx.recv().unwrap();
    });
    while tx.send(1).is_ok() {}
}
```

As expected, `rx.recv().unwrap()` executes, sees the channel has no data, and suspends. It does this by parking using the generic parker interface which involves waiting on a condvar. It successfully sends the request to `wait()`.

Afterwards, `tx.send(1)` is executed. Because of the current quantum, this is allowed to execute quite a lot. In fact, it executes over two million times, which easily eats up the eight megabytes of RAM. The other thread does not execute because it is waiting on a condvar.

What appears to happen is the `rx` thread registers itself as a `selector` using `self.receivers.register(oper, cx)`, but the `tx` thread seems to query the `observers`.

Furthermore, the `observers` field never appears to actually be used.

### Meta

This is based off Rust nightly 496928c274a438558fe165c110c57a7cb583b308

**Comment by xobs** [comment]
After understanding the `mpmc` crate a bit more (and after help from some people on Zulip), I've discovered that `observers` is, indeed, vestigial and was a red herring.

However, I still believe this test is unsound on single-core systems. In fact, I think three tests are unsound on single-core systems, particularly those with limited RAM:

* test_recv_try_iter
* port_gone_concurrent
* port_gone_concurrent_shared

The problem with `port_gone_concurrent` comes down to whether the spawned thread runs first or last, and the behaviour of the scheduler. The following sequence of events works:

1. The spawned thread executes first
2. It attempts to receive data, fails, and parks itself
3. The parent thread runs and sends data into the channel
4. The child thread is signalled and is immediately switched to
5. The child thread receives data and terminates
6. The parent thread tries to send data again, fails, and terminates

Any deviation from this will result in an unbounded execution waiting for a context switch. For example:

1. The parent thread executes first
2. The parent thread executes `while tx.send(1).is_ok() {}` until it has exhausted all memory

Or if the child thread executes first:

1. The child thread attempts to read from the channel
2. The child thread is parked
3. The parent thread executes `while tx.send(1).is_ok() {}` which unparks the child thread but does not deschedule the parent thread
4. Memory is exhausted

There are three solutions to this that I can see:

1. Run this test on a system that has multiple cores, where both threads can run simultaneously
2. Have enough RAM that the sender can allocate an infinitely large buffer before it gets descheduled
3. Add `crate::thread::yield_now()` every few thousand iterations, allowing the other threads time to run

**Comment by xobs** [comment]
This test was useful in discovering a resource-starvation issue. It may be a problem with devices that have even less RAM, but after adding optimizations to the kernel to force context switches on condvar unlocks, this problem has gone away.
